runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java

```diff
@@ -0,0 +1,88 @@
+package org.apache.beam.runners.flink;
+
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import io.grpc.stub.StreamObserver;
+import javax.annotation.Nullable;
+import org.apache.beam.model.jobmanagement.v1.JobApi.JobMessage;
+import org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum;
+import org.apache.beam.runners.fnexecution.jobsubmission.JobInvocation;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Invocation of a Flink Job via {@link FlinkRunner}.
+ */
+public class FlinkJobInvocation implements JobInvocation {
```

Comment:
Just throwing out some ideas:

I managed to make some changes to the Flink `ClusterClient` interface. This should make it easier to implement the methods of this: https://github.com/apache/flink/blob/10d52f268db3eda7ee1511ea30afb9a982644148/flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java. Notable are:
 - `submitJob(JobGraph, ClassLoader)`, which is non-blocking if you all `setDetached(true)` before
 - `cancel(JobID)`
 - `getJobStatus(JobID)`

A `JobGraph` will have a random `JobID`, you can get it from the `JobGraph` via `getJobID()`. For streaming you can get the `JobGraph` via `StreamExecutionEnvironment.getStreamGraph().getJobGraph()`. For batch it's a bit more involved but a starting point is this: https://github.com/apache/flink/blob/10d52f268db3eda7ee1511ea30afb9a982644148/flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java#L890.

In general those APIs are somewhat internal and they are also somewhat interesting... 😉